Reusable Workflows for building and deployment

bunbry opened the pull request:
Reusable Workflows for building and deploying AKTIN maven projects, debian packages, and docker images.

bunbry committed:
ci: add get-newest-artifact composite action

bunbry committed:
WIP ci: add build & deploy workflows for AKTIN Maven, Debian, and Docker projects

bunbry committed:
WIP fix: workflows must be rooted in .github/workflows directory

bunbry committed:
feat: set "downloads" cache key based on versions file

bunbry committed:
refactor!: remove steps related to i2b2 download

i2b2 docker and debian build.sh now integrate the downloads fo i2b2 files.

BREAKING CHANGE: reusable workflow input "download-i2b2" removed

bunbry committed:
refactor: remove passphrase authentication for SSH

Both passphrase and private key authentication were implemented but only private key authentication will be used.

bunbry committed:
refactor: update remote address for debian deployment

bunbry committed:
fix: actually make use of cached AKTIN repo public key

bunbry committed:
refactor!: rename input for debian-build.yml

For consistency we use the same term as AKTIN debian build.sh which use "PACKAGE_NAME".

BREAKING CHANGE: reusable workflow input "name-base" renamed to "package-name"

bunbry committed:
refactor: uncomment pre-testing installation steps

bunbry committed:
refactor: add REPREPRO_HOME value

bunbry committed:
docs: correct workflow directory in example

bunbry committed:
fix(debian build): remove build script version parameter

The build.sh now retrieves the version from a versions file instead of from script parameter.

bunbry committed:
refactor(debian build): remove installation and testing steps

The debian testing will be handled in the repositories of the debian packages.

bunbry committed:
refactor(debian build): exclude lines starting with # from cache name

The versions file now contains comments which use # as first character of the line.

bunbry committed:
refactor(debian build): use version from versions file instead of git tag

The build.sh now retrieves the version from a versions file instead of from script parameter. This version contains ~ to indicate unstable releases as per debian versioning guidelines. The ~ are not contained in the git tag, which are the previous source of the version.

bunbry committed:
refactor(debian deploy): remove manipulation of the distributions file

bunbry committed:
refactor(docker): remove docker build and deploy workflows

The docker build and deploy workflows are moved into the docker repository. They are to specific to be reusable especially because now we only create DWH images and not i2b2 and DWH images.

bunbry committed:
fix(debian build): move working-directory from job to step

The actions "cache" and "upload-artifact" are sensitive to the working directory of the job, which was not considered consistently.
Due to previous refactoring, a job wide working directory might be less readable while providing little benefit.

bunbry committed:
feat(debian deploy)!: determine deployed package from inputs

Avoiding glob (*.deb) in package name.

BREAKING CHANGE: new required inputs package-name and package-version

bunbry committed:
feat(debian build): package version as workflow output

Avoiding glob (*.deb) in package name in other workflows.

bunbry committed:
refactor(debian build): always get package version from versions file

Discussions about how versions should be formed have been postponed.

bunbry committed:
fix(debian deploy): remove "needs" section from deploy job

This line was accidentally left in from when the build and the deploy job were in the same workflow file.

bunbry committed:
docs(debian): fix some spelling errors and rephrase description of workflow inputs and names

bunbry committed:
feat(debian build): fail if artifact contains no files

akomii committed:
updated readme

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,61 @@
+        type: string
+    outputs:
+      package-version:
+        description: 'Version of the debian package'
```
Version of the **built** debian package

akomii replied:
Version of the **built** debian package

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,61 @@
+#        if: ! startsWith(github.ref, 'refs/tags/v')
+#        run: echo "VERSION=0-$GITHUB_REF_NAME" >> $GITHUB_ENV
+
+      - uses: actions/checkout@v4
```
Shouldnt this be the very first step?

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,61 @@
+        run: |
+          STR=$(grep "PACKAGE_VERSION=" versions)
+          VERSION=${STR#PACKAGE_VERSION=}
+          echo "VERSION=$VERSION" >> $GITHUB_ENV
+          echo "version=$VERSION" >> $GITHUB_OUTPUT
```
Why is version once in caps and once in lowercase?

bunbry replied:
Shell environment variables are conventionally capitalized, while GitHub workflow variables are lowercase.

`echo "VERSION=$VERSION" >> $GITHUB_ENV` adds the environment variable `VERSION` to a job wide context (usable by other steps) using the previously locally defined environment variable of the same name.
`echo "version=$VERSION" >> $GITHUB_OUTPUT` adds the output variable `version` to the output of the step, still using the locally defined variable.

akomii replied:
Alrighty

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,61 @@
+          echo "version=$VERSION" >> $GITHUB_OUTPUT
+        working-directory: src/resources
+
+      - name: Prepare Version Variable - Testing
+        if: ! startsWith(github.ref, 'refs/tags/v')
+        run: echo "VERSION=0-$GITHUB_REF_NAME" >> $GITHUB_ENV
+#      - name: Prepare Version Variable - Testing
+#        if: ! startsWith(github.ref, 'refs/tags/v')
+#        run: echo "VERSION=0-$GITHUB_REF_NAME" >> $GITHUB_ENV
```
It is unclear why this happens? Is this still a test? Also, no commented out code should be in the PR

bunbry replied:
This step was a placeholder and left in accidentally (and is now removed). However, we should consider adding unique build numbers (generated in the CI pipeline) to the version for development builds.

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,61 @@
+
+      - name: Build Debian
+        run: build.sh
+        working-directory: src/debian
```
Verify, that build.sh is executable befor executing it, i.e. by calling:
> chmod +x src/debian/build.sh
before this step

akomii commented on .github/workflows/debian-deploy.yml:
```diff
@@ -0,0 +1,52 @@
+        type: string
+      codename:
+        description: 'OS codename'
+        default: 'noble'
```
We downgraded the Release to Ubuntu 22. So the default codename is now "jammy"

akomii commented on .github/workflows/debian-deploy.yml:
```diff
@@ -0,0 +1,52 @@
+    runs-on: ubuntu-latest
+    if: ${{ startsWith(github.ref, 'refs/tags/v') }}
+    env:
+      REPREPRO_HOME: /software/repo/org/apt   # local directory containing conf, db, dists, and pool directories
```
Not the home of reprepro, but of our repository. This is (to some level) confidential information. Better make it an input with default value

akomii commented on .github/workflows/debian-deploy.yml:
```diff
@@ -0,0 +1,52 @@
+          ssh-private-key: ${{ secrets.SSH_PRIVATE_KEY }}
+
+      - name: Deploy Debian Package
+        run: scp $PACKAGE "${{ secrets.SSH_USER }}@a2f8a.netcupt.net:$REPREPRO_HOME/"
```
Hostname is confidential and should be a secret

bunbry replied:
This is the same hostname which is also published in the pom.xml on the master branch of the [aktin/aktin](https://github.com/aktin/aktin) repository. See related comment for maven CI: https://github.com/aktin/aktin-github-scripts/pull/2#discussion_r1925575146

akomii commented on .github/workflows/debian-deploy.yml:
```diff
@@ -0,0 +1,52 @@
+        run: scp $PACKAGE "${{ secrets.SSH_USER }}@a2f8a.netcupt.net:$REPREPRO_HOME/"
+
+      - name: Import Debian Packages
+        run: ssh ${{ secrets.SSH_USER }}@a2f8a.netcupt.net "reprepro -b '$REPREPRO_HOME' includedeb ${{ inputs.codename }} $PACKAGE"
```
Consider that the package needs to be signed with a gpg key. Chatgpt recommended this:

```
secrets:
  GPG_PRIVATE_KEY:
    description: 'GPG private key in ASCII armor format'
    required: true
  GPG_PASSPHRASE:
    description: 'Passphrase for the GPG key'
    required: true

steps:
  # ... previous steps ...

  - name: Import GPG Key
    run: |
      echo "${{ secrets.GPG_PRIVATE_KEY }}" | gpg --batch --import
      gpg --list-secret-keys

  - name: Deploy and Sign Package
    env:
      GPG_PASS: ${{ secrets.GPG_PASSPHRASE }}
    run: |
      scp "$PACKAGE" "${{ secrets.SSH_USER }}@${{ inputs.repo-host }}:${{ inputs.repo-path }}/" || exit 1
      ssh "${{ secrets.SSH_USER }}@${{ inputs.repo-host }}" \
        "GNUPGHOME=/path/to/gpg \
         GPG_TTY=\$(tty) \
         echo '$GPG_PASS' | \
         reprepro -b '${{ inputs.repo-path }}' --ask-passphrase includedeb ${{ inputs.codename }} $PACKAGE" || exit 1
```

bunbry replied:
The idea for the existing code was "the key is always loaded in the remote gpg-agent and does not require further passphrase input". Further testing shows that the gpg-agent does not cache passphrases between sessions, so we need to input it anyway, but it keeps the encrypted GPG key loaded. 
The passphrase cannot be handed as an input to the `reprepro` command in a non-interactive way, it needs to be preloaded into the gpg-agent. For that, `allow-preset-passphrase` must be set (once) in the agents' config file of the user.

Do you have any concerns regarding the `allow-preset-passphrase` setting?
Should the GPG key be load/unloaded for every session (the preset password will be removed either way)?

akomii reviewed (commented):
maven-build-deploy.yml not reviewed.

As communicated, remove
- debian-deploy-gh
- debian-deploy-override
- get-newest-artifact

bunbry committed:
refactor: remove unused workflow versions for deployment of debian files

bunbry committed:
refactor: remove unused action "get-newest-artifact"

bunbry committed:
fix(debian build): use checkout action as first step / before version preparation

With the new way of retrieving the versions (from version files instead for git tag) checkout needs to happen prior to version preparation.

bunbry committed:
feat(debian build): ensure that build.sh is executable

bunbry committed:
refactor(debian build): remove obsolete version generation step for non-deployment cases

bunbry committed:
docs(debian build): rephrase description for output "package-version"

bunbry committed:
refactor(debian deploy): downgrade default OS to Ubuntu 22.04

bunbry committed:
feat(debian deploy): hostname of apt repo server as secret

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,83 @@
+        description: 'Whether to install r-base and r-cran-tidyverse before testing'
+        default: false
+        required: false
+        type: boolean
```
This is a special and uncommon use-case that should better be handled by the parent workflows directly

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,83 @@
+          java-version: ${{ inputs.java-version }}
+          distribution: ${{ inputs.java-distribution }}
+          cache: maven
+          server-id: aktinupload.software.magdeburg  # value of the distributionManagement/repository/id field of the pom.xml
```
server-id should be a secret

bunbry replied:
The server-id is part of the pom. Its value is arbitrary (see [setup-java maven options](https://github.com/actions/setup-java/tree/v4?tab=readme-ov-file#maven-options)), and is not the hostname of a server (the hostname is also mentioned in the pom).
The hostname could be excised from the pom and inserted via a secret wrapped in an environment variable during CI. 
**Note:** The hostname is already part of the master branch for the [aktin/aktin](https://github.com/aktin/aktin) repository. A hostname change or force pushes to rewrite history would be required.

akomii replied:
Better to declare this as an env in L40 then

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,83 @@
+
+      - name: Install APT Dependencies
+        if: inputs.install-r
+        run: sudo apt install -y r-base r-cran-tidyverse
```
Maybe generify this step to open it  up for future requirements, like:

```
install-apt:
    description: 'Space-separated list of APT packages to install'
    default: ''
    required: false
    type: string
 ```
  and 
  ```
- name: Install APT Dependencies
    if: inputs.install-apt != ''
    run: sudo apt update && sudo apt install -y ${{ inputs.install-apt }}
```

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,83 @@
+        if: startsWith(github.ref, 'refs/tags/v')
+#        run: mvn deploy -Drevision=${GITHUB_REF_NAME#v} -Dchangelist=
+        run: |
+          mvn versions:set -DnewVersion=${GITHUB_REF_NAME#v}
```
`-DprocessAllModules=true` to update all modules recursively

bunbry replied:
The `mvn versions:set -DnewVersion=${GITHUB_REF_NAME#v}` is only applied to the module of the repository which triggers this workflow. No other modules are checked out during the workflow.

bunbry replied:
I tested the behaviour of `versions:set` with and without `-DprocessAllModules=true`: When set to `true`, the version of a module is always overwritten, otherwise the version is only overwritten if it is the same as the old version of the parent.

akomii replied:
Packages should that are not at the top hierarchy (i.e. `dwh-query`) should always share the same version between parent (`dwh-query`) and child (i.e. `generic-file-importer`)

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,83 @@
+        if: startsWith(github.ref, 'refs/tags/v')
+#        run: mvn deploy -Drevision=${GITHUB_REF_NAME#v} -Dchangelist=
+        run: |
+          mvn versions:set -DnewVersion=${GITHUB_REF_NAME#v}
+          mvn deploy
```
Add also here a bump to the next version, i.e. 

> pom.xml has version 0.19rc1-SNAPSHOT
> User tags main branch with 0.19
> Workflow releases the artifact as 0.19
> Workflow updates pom.xml to 0.20-SNAPSHOT

Chatgpt recommends:
```
  run: |
    RELEASE_VERSION=${GITHUB_REF_NAME#v}
    NEXT_VERSION=$(echo $RELEASE_VERSION | awk -F. '{print $1"."$2+1"-SNAPSHOT"}')

    # Set release version
    mvn versions:set -DnewVersion=$RELEASE_VERSION
    mvn deploy

    # Bump to next snapshot
    mvn versions:set -DnewVersion=$NEXT_VERSION
    mvn versions:commit

    # Push updated version back to the repo
    git config --global user.name "github-actions"
    git config --global user.email "actions@github.com"
    git commit -am "Bump version to $NEXT_VERSION"
    git push
```

bunbry replied:
The recommended way of handling versions with CI is to set the project version in the pom files to something which is clearly not a normal version and inserting real versions during CI. The [CI Friendly Versions](https://maven.apache.org/guides/mini/guide-maven-ci-friendly.html) placeholders remove the need for manual propagation of the parent pom versions. 
Use of the Maven Release plugin or any other method which creates its own commits is discouraged.

bunbry committed:
refactor(debian deploy): add path to repository as an input

Having the path to the repository (formerly REPREPRO_HOME) as an input is more generic.

bunbry committed:
refactor(maven)!: change "install-r" workflow to input for generic apt dependencies

Make workflow more generic.

BREAKING CHANGE: removes reusable workflow input "install-r"

bunbry committed:
feat(maven): add workflow secret "AKTIN_MAVEN_REPOSITORY"

Allows the keep the hostname of the maven repository confidential.
Note: It requires that the "MAVEN_REPOSITORY" environment variable is used as value of the distributionManagement/repository/id field of the pom.xml, or it will have no effect.

bunbry committed:
refactor(debian deploy): remove unused environment variable REPREPRO_HOME

Accidentally not removed in 909c0183.

bunbry committed:
fix(debian deploy)!: preset and remove GPG passphrase in remote gpg-agent

BREAKING CHANGE: adds required secrets "GPG_PASSPHRASE" and "GPG_KEYGRIP"

bunbry committed:
fix(debian deploy): add " around environment variable

bunbry committed:
docs(debian deploy): rephrase step names

bunbry commented:
We should have a discussion about how the versions of projects and their modules are intended to relate to each other. I think there is some potential for simplifying the poms, which could give some quality of life improvements even without applying CI friendly versions. 
Related comments: https://github.com/aktin/aktin-github-scripts/pull/2#discussion_r1925420274, https://github.com/aktin/aktin-github-scripts/pull/2#discussion_r1926861656

Also, I noticed that I misapplied the use of CI friendly version placeholders with regard to the aktin pom and its child poms because of the odd relationship between the projects. The corresponding pull requests have been marked as draft: aktin/aktin#4, aktin/dwh-admin#18, aktin/dwh-api#1, aktin/dwh-db#4, aktin/dwh-import#7, aktin/dwh-j2ee#4, aktin/dwh-prefs#6, aktin/dwh-query#12, aktin/dwh-setup#1, aktin/r-script#2

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,57 @@
+          echo "version=$VERSION" >> $GITHUB_OUTPUT
+        working-directory: src/resources
+
+      - name: Generate Cache Name
+        run: echo "CACHE=${sort versions | awk '!/(^[[:space:]]*$|PACKAGE_VERSION=|#)/' ORS='--'}" >> $GITHUB_ENV
```
Isnt here a `working-directory` missing or is it pulled automatically from the previous step?

akomii commented on .github/workflows/debian-build.yml:
```diff
@@ -0,0 +1,57 @@
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ inputs.artifact-name }}
+          path: src/debian/build/${{ inputs.package-name }}_$VERSION.deb
```
I would suggest to use here `${{ env.VERSION }}` for consistency

akomii commented on .github/workflows/debian-deploy.yml:
A cleanup is missing. We should use an `incoming` configuration, i.e.


> Die Konfigurations-Datei incoming bestimmt die Regeln und Ordner beim Upload von neuen Paketen ins reprepro Repository. Die Konfigurations-Datei incoming befindet sich im /conf Ordner des verwendeten reprepro Repos.
```
:~$ vi packages/conf/incoming
Name: incoming
IncomingDir: /home/repository/incoming
Allow: precise
Cleanup: on_deny on_error
Tempdir: /home/repository/incoming_tmp
LogDir: /home/repository/incoming_log
```

akomii replied:
Best to add a new input var called `incoming-path`

akomii replied:
Postponed because of requirements of `changes` file

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,86 @@
+  workflow_call:
+    inputs:
+      java-version:
+        description: 'Java version number'
```
Java LTS version (8, 11, 17, 21)

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,86 @@
+          server-username: MAVEN_USERNAME  # env variable for username in deploy
+          server-password: MAVEN_PASSPHRASE  # env variable for token/password in deploy
+          gpg-private-key: ${{ secrets.AKTIN_MAVEN_GPG_PRIVATE_KEY }}  # value of the GPG private key to import
+          gpg-passphrase: MAVEN_GPG_PASSPHRASE  # env variable for GPG private key passphrase
```
${{ secrets.AKTIN_MAVEN_GPG_PASSPHRASE }}

bunbry replied:
Here we set the name of the environment variable, not its value (note that the comments start with either "env variable" or "value"). The secret is used in the "Deploy" step.

akomii commented on .github/workflows/maven-build-deploy.yml:
```diff
@@ -0,0 +1,86 @@
+#        run: mvn deploy -Drevision=${GITHUB_REF_NAME#v} -Dchangelist=
+        run: |
+          mvn versions:set -DnewVersion=${GITHUB_REF_NAME#v}
+          mvn deploy
```
Add here `--no-transfer-progress` to suppress download progress bars

bunbry committed:
feat(debian deploy): list available packages after deployment

bunbry committed:
docs(maven): java version description LTS only

bunbry committed:
feat(maven): suppress download progress during build and deploy

bunbry committed:
fix(debian build): add working directory to cache generation step

bunbry committed:
style(debian build): use ${{env.VERSION}} not $VERSION

More consistent look.

bunbry committed:
refactor(maven) remove commented out lines

bunbry committed:
style: consistent use of free lines and empty last lines

bunbry committed:
feat(debian deploy): add cleanup on remote

bunbry committed:
fix(debian deploy): explicit path to package
